File: Plantemon/app/src/main/java/com/g4ng/model/Plant.java
package com.g4ng.model;

import com.g4ng.logic.Move;

import java.io.Serializable;
import java.util.UUID;
import java.util.List;
import java.util.ArrayList;
import java.util.Date;

public class Plant implements Serializable {
    private final UUID id;
    private final String name;
    private final int maxHealth;
    private int currentHealth;
    private int speed;
    private final List<Move> moves;
    private final String spritePath;
    private Date scanDateTime;

    // New metadata fields from API
    private List<String> commonNames;
    private String description;
    private String taxonomy;
    private String bestLightCondition;
    private String bestSoilType;
    private String commonUses;
    private String culturalSignificance;
    private String toxicity;
    private String bestWatering;

    public Plant(String name, int speed, String spritePath) {
        this.id = UUID.randomUUID();
        this.name = name;
        this.maxHealth = 100;
        this.currentHealth = maxHealth;
        this.speed = speed;
        this.spritePath = spritePath;
        this.moves = new ArrayList<>();
        this.scanDateTime = new Date(); // Default to now
    }
    /**
     * Copy constructor for deep copying a Plant object.
     */
    public Plant(Plant other) {
        this.id = other.id;
        this.name = other.name;
        this.spritePath = other.spritePath;

        this.maxHealth = other.maxHealth;
        this.currentHealth = other.currentHealth;
        this.speed = other.speed;

        this.moves = other.moves;

        if (other.scanDateTime != null) {
            this.scanDateTime = new Date(other.scanDateTime.getTime());
        }
        if (other.commonNames != null) {
            this.commonNames = new ArrayList<>(other.commonNames);
        } else {
            this.commonNames = null;
        }
        this.description = other.description;
        this.taxonomy = other.taxonomy;
        this.bestLightCondition = other.bestLightCondition;
        this.bestSoilType = other.bestSoilType;
        this.commonUses = other.commonUses;
        this.culturalSignificance = other.culturalSignificance;
        this.toxicity = other.toxicity;
        this.bestWatering = other.bestWatering;
    }

    // Setters
    public void setCommonNames(List<String> commonNames) { this.commonNames = commonNames; }
    public void setDescription(String description) { this.description = description; }
    public void setTaxonomy(String taxonomy) { this.taxonomy = taxonomy; }
    public void setBestLightCondition(String bestLightCondition) { this.bestLightCondition = bestLightCondition; }
    public void setBestSoilType(String bestSoilType) { this.bestSoilType = bestSoilType; }
    public void setCommonUses(String commonUses) { this.commonUses = commonUses; }
    public void setCulturalSignificance(String culturalSignificance) { this.culturalSignificance = culturalSignificance; }
    public void setToxicity(String toxicity) { this.toxicity = toxicity; }
    public void setBestWatering(String bestWatering) { this.bestWatering = bestWatering; }

    // Getters
    public UUID getId() { return id; }
    public String getName() { return name; }
    public int getSpeed() { return speed; }
    public List<Move> getMoves() { return moves; }
    public int getMaxHealth() { return maxHealth; }
    public int getCurrentHealth() { return currentHealth; }
    public void setCurrentHealth(int health) { 
        this.currentHealth = Math.min(health, this.maxHealth); 
    }
    public void takeDamage(int amount) {
        this.currentHealth = Math.max(this.currentHealth - amount, 0);
    }
    public boolean isDead() {
        return currentHealth == 0;
    }

    public String getDescription() { return description; }
    public String getBestLightCondition() { return bestLightCondition; }
    public String getBestSoilType() { return bestSoilType; }
    public String getCommonUses() { return commonUses; }
    public String getToxicity() { return toxicity; }
    public String getBestWatering() { return bestWatering; }
    public Date getScanDateTime() { return scanDateTime; }

    public String getSpritePath() { return spritePath; }

    public void addMove(Move move) { this.moves.add(move); }

    public String getTaxonomy() { return taxonomy; 
    }

    public void setScanDateTime(Date scanDateTime) {
        this.scanDateTime = scanDateTime;
    }
}

File: plantemon-web/src/lib/domain/plant.ts
import type { Move } from "./move";

/** Metadata copied off the Plant.id response. Port of the fields on model/Plant.java. */
export interface PlantMetadata {
  commonNames?: string[] | null;
  description?: string | null;
  taxonomy?: string | null;
  bestLightCondition?: string | null;
  bestSoilType?: string | null;
  commonUses?: string | null;
  culturalSignificance?: string | null;
  toxicity?: string | null;
  bestWatering?: string | null;
}

/**
 * Port of model/Plant.java.
 *
 * maxHealth is fixed at 100 for every plant, as in the original.
 */
export class Plant {
  readonly id: string;
  readonly name: string;
  readonly maxHealth: number = 100;
  readonly spritePath: string | null;

  private currentHealth: number;
  private speed: number;
  private readonly moves: Move[] = [];
  private scanDateTime: Date;

  commonNames: string[] | null = null;
  description: string | null = null;
  taxonomy: string | null = null;
  bestLightCondition: string | null = null;
  bestSoilType: string | null = null;
  commonUses: string | null = null;
  culturalSignificance: string | null = null;
  toxicity: string | null = null;
  bestWatering: string | null = null;

  constructor(name: string, speed: number, spritePath: string | null, id?: string) {
    this.id = id ?? crypto.randomUUID();
    this.name = name;
    this.currentHealth = this.maxHealth;
    this.speed = speed;
    this.spritePath = spritePath;
    this.scanDateTime = new Date();
  }

  /**
   * Deep-ish copy, matching the Java copy constructor — which deliberately
   * shares the `moves` list reference rather than cloning it.
   */
  static copyOf(other: Plant): Plant {
    const copy = new Plant(other.name, other.speed, other.spritePath, other.id);
    copy.currentHealth = other.currentHealth;
    copy.scanDateTime = new Date(other.scanDateTime.getTime());
    copy.moves.push(...other.moves);
    copy.commonNames = other.commonNames ? [...other.commonNames] : null;
    copy.description = other.description;
    copy.taxonomy = other.taxonomy;
    copy.bestLightCondition = other.bestLightCondition;
    copy.bestSoilType = other.bestSoilType;
    copy.commonUses = other.commonUses;
    copy.culturalSignificance = other.culturalSignificance;
    copy.toxicity = other.toxicity;
    copy.bestWatering = other.bestWatering;
    return copy;
  }

  getName(): string {
    return this.name;
  }

  getSpeed(): number {
    return this.speed;
  }

  setSpeed(speed: number): void {
    this.speed = speed;
  }

  getMoves(): Move[] {
    return this.moves;
  }

  addMove(move: Move): void {
    this.moves.push(move);
  }

  getMaxHealth(): number {
    return this.maxHealth;
  }

  getCurrentHealth(): number {
    return this.currentHealth;
  }

  /** Caps at maxHealth. Note the original applies no lower bound here. */
  setCurrentHealth(health: number): void {
    this.currentHealth = Math.min(health, this.maxHealth);
  }

  takeDamage(amount: number): void {
    this.currentHealth = Math.max(this.currentHealth - amount, 0);
  }

  isDead(): boolean {
    return this.currentHealth === 0;
  }

  getScanDateTime(): Date {
    return this.scanDateTime;
  }

  setScanDateTime(date: Date): void {
    this.scanDateTime = date;
  }

  applyMetadata(metadata: PlantMetadata): void {
    this.commonNames = metadata.commonNames ?? null;
    this.description = metadata.description ?? null;
    this.taxonomy = metadata.taxonomy ?? null;
    this.bestLightCondition = metadata.bestLightCondition ?? null;
    this.bestSoilType = metadata.bestSoilType ?? null;
    this.commonUses = metadata.commonUses ?? null;
    this.culturalSignificance = metadata.culturalSignificance ?? null;
    this.toxicity = metadata.toxicity ?? null;
    this.bestWatering = metadata.bestWatering ?? null;
  }
}
